scripts/recon/train_unrolled.py

```diff
@@ -0,0 +1,313 @@
+    else:
+        raise ValueError(f"Unsuported optimizer : {config.optimizer.type}")
+    metrics = {
+        "LOSS": [],
+        "MSE": [],
+        "MAE": [],
+        "LPIPS": [],
+        "PSNR": [],
+        "SSIM": [],
+        "ReconstructionError": [],
+        "n_iter": n_iter,
+        "algorithm": config.reconstruction.method,
+    }
+
+    # Backward hook that detect NAN in the gradient and print the layer weights
+    def detect_nan(grad):
+        if torch.isnan(grad).any():
+            print(grad)
+            for name, param in recon.named_parameters():
+                print(name, param)
+            raise ValueError("Gradient is NaN")
+        return grad
+
+    for param in recon.parameters():
+        param.register_hook(detect_nan)
+
+    # Training loop
+    for epoch in range(config.training.epoch):
+        print(f"Epoch {epoch}")
+        mean_loss = 0.0
+        i = 1.0
+        pbar = tqdm(data_loader)
+        for X, y in pbar:
+            # send to device
+            X = X.to(device)
+            y = y.to(device)
+            if X.shape[3] == 3:
+                X = X
+                y = y
+
+            y_pred = recon.batch_call(X.to(device))
+            # normalizing each output
+            y_pred_max = torch.amax(y_pred, dim=(-1, -2, -3), keepdim=True)
+            y_pred = y_pred / y_pred_max
+
+            # normalizing y
+            y = y.to(device)
+            y_max = torch.amax(y, dim=(-1, -2, -3), keepdim=True)
+            y = y / y_max
+
+            if i % disp == 1 and config.display.plot:
+                img_pred = y_pred[0, 0].cpu().detach().numpy()
+                img_truth = y[0, 0].cpu().detach().numpy()
+
+                plt.imshow(img_pred)
+                plt.savefig(f"y_pred_{i-1}.png")
+                plt.imshow(img_truth)
+                plt.savefig(f"y_{i-1}.png")
+
+            optimizer.zero_grad(set_to_none=True)
+            # convert to CHW for loss and remove depth
+            y_pred = y_pred.reshape(-1, *y_pred.shape[-3:]).movedim(-1, -3)
+            y = y.reshape(-1, *y.shape[-3:]).movedim(-1, -3)
+
+            loss_v = Loss(y_pred, y)
+            if config.lpips:
+                loss_v = loss_v + config.lpips * torch.mean(loss_lpips(y_pred, y))
+            loss_v.backward()
+            torch.nn.utils.clip_grad_norm_(recon.parameters(), 1.0)
+            optimizer.step()
+
+            mean_loss += (loss_v.item() - mean_loss) * (1 / i)
+
+            pbar.set_description(f"loss : {mean_loss}")
+            i += 1
+
+        # benchmarking
+        current_metrics = benchmark(recon, benchmark_dataset, batchsize=10)
+        # update metrics with current metrics
+        metrics["LOSS"].append(mean_loss)
+        for key in current_metrics:
+            metrics[key].append(current_metrics[key])
+
+    print(f"Train time : {time.time() - start_time} s")
+
+    # save dictionary metrics to file with json
+    try:
+        import json
```

Comment:
import earlier? would be a shame to run the whole training for it to error here :sweat_smile: